cpp/include/raft/spatial/knn/detail/ann_utils.cuh

```diff
@@ -0,0 +1,491 @@
+ *
+ * @param[in] n_rows
+ * @param[in] n_cols
+ * @param[inout] a device pointer to a row-major matrix [n_rows, n_cols]
+ * @param[in] d device pointer to a vector of divisors [n_rows]
+ */
+inline void divide_along_rows(
+  uint32_t n_rows, uint32_t n_cols, float* a, const uint32_t* d, rmm::cuda_stream_view stream)
+{
+  dim3 threads(128, 1, 1);
+  dim3 blocks(
+    ceildiv<uint64_t>(static_cast<uint64_t>(n_rows) * static_cast<uint64_t>(n_cols), threads.x),
+    1,
+    1);
+  divide_along_rows_kernel<<<blocks, threads, 0, stream>>>(n_rows, n_cols, a, d);
+}
+
+template <typename T>
+__global__ void outer_add_kernel(const T* a, uint32_t len_a, const T* b, uint32_t len_b, T* c)
+{
+  uint64_t gid = threadIdx.x + blockDim.x * blockIdx.x;
+  uint64_t i   = gid / len_b;
+  uint64_t j   = gid % len_b;
+  if (i >= len_a) return;
+  c[gid] = (a == nullptr ? T(0) : a[i]) + (b == nullptr ? T(0) : b[j]);
+}
+
+/**
+ * @brief Fill matrix `c` with all combinations of sums of vectors `a` and `b`.
+ *
+ * NB: device-only function
+ *
+ * @tparam T element type
+ *
+ * @param[in] a device pointer to a vector [len_a]
+ * @param len_a number of elements in `a`
+ * @param[in] b device pointer to a vector [len_b]
+ * @param len_b number of elements in `b`
+ * @param[out] c row-major matrix [len_a, len_b]
+ * @param stream
+ */
+template <typename T>
+void outer_add(
+  const T* a, uint32_t len_a, const T* b, uint32_t len_b, T* c, rmm::cuda_stream_view stream)
+{
+  dim3 threads(128, 1, 1);
+  dim3 blocks(
+    ceildiv<uint64_t>(static_cast<uint64_t>(len_a) * static_cast<uint64_t>(len_b), threads.x),
+    1,
+    1);
+  outer_add_kernel<<<blocks, threads, 0, stream>>>(a, len_a, b, len_b, c);
+}
+
+template <typename T, typename S>
+__global__ void copy_selected_kernel(uint32_t n_rows,
+                                     uint32_t n_cols,
+                                     const S* src,
+                                     const uint32_t* row_ids,
+                                     uint32_t ld_src,
+                                     T* dst,
+                                     uint32_t ld_dst)
+{
+  uint64_t gid   = threadIdx.x + blockDim.x * blockIdx.x;
+  uint64_t j     = gid % n_cols;
+  uint64_t i_dst = gid / n_cols;
+  if (i_dst >= n_rows) return;
+  uint64_t i_src          = row_ids[i_dst];
+  dst[ld_dst * i_dst + j] = mapping<T>{}(src[ld_src * i_src + j]);
+}
+
+/**
+ * @brief Copy selected rows of a matrix while mapping the data from the source to the target
+ * type.
+ *
+ * @tparam T target type
+ * @tparam S source type
+ *
+ * @param n_rows
+ * @param n_cols
+ * @param[in] src input matrix [..., ld_src]
+ * @param[in] row_ids selection of rows to be copied [n_rows]
+ * @param ld_src number of cols in the input (ld_src >= n_cols)
+ * @param[out] dst output matrix [n_rows, ld_dst]
+ * @param ld_dst number of cols in the output (ld_dst >= n_cols)
+ * @param stream
+ */
+template <typename T, typename S>
+void copy_selected(uint32_t n_rows,
```

Comment:
It may have different source and target types. The mapping function is hardcoded, so if we want to move this to the global namespace, we should make it into a lambda.